src/engine/TransitivePathBase.cpp

```diff
@@ -36,29 +37,54 @@ TransitivePathBase::TransitivePathBase(
   AD_CORRECTNESS_CHECK(qec != nullptr);
   AD_CORRECTNESS_CHECK(subtree_);
   if (lhs_.isVariable()) {
-    variableColumns_[std::get<Variable>(lhs_.value_)] =
-        makeAlwaysDefinedColumn(0);
+    variableColumns_[lhs_.value_.getVariable()] = makeAlwaysDefinedColumn(0);
   }
   if (rhs_.isVariable()) {
-    variableColumns_[std::get<Variable>(rhs_.value_)] =
-        makeAlwaysDefinedColumn(1);
+    variableColumns_[rhs_.value_.getVariable()] = makeAlwaysDefinedColumn(1);
   }
-  if (minDist_ == 0 && lhs_.isUnboundVariable() && rhs_.isUnboundVariable()) {
-    emptyPathBound_ = true;
-    lhs_.treeAndCol_.emplace(makeEmptyPathSide(qec, std::move(activeGraphs)),
-                             0);
+  if (minDist_ == 0) {
+    auto& startingSide = decideDirection().first;
+    // If we have hardcoded differing values left and right, we can increase the
+    // minimum distance to 1
+    if (!lhs_.isVariable() && !rhs_.isVariable() &&
+        lhs_.value_ != rhs_.value_) {
+      minDist_ = 1;
+    } else if (lhs_.isUnboundVariable() && rhs_.isUnboundVariable()) {
+      boundVariableIsForEmptyPath_ = true;
+      lhs_.treeAndCol_.emplace(makeEmptyPathSide(qec, std::move(activeGraphs)),
+                               0);
+    } else if (!startingSide.isVariable()) {
+      startingSide.treeAndCol_.emplace(
+          joinWithIndexScan(qec, std::move(activeGraphs), startingSide.value_),
+          0);
```

Comment:
I would stall this change until we have proper prefiltering, because this will make many 
cases very very expensive. 
Can you for now e.g. hide it behind a runtime parameter that defaults to the old behavior?

Reply:
In fact I will check whether you have implemented it efficiently.